core/src/main/java/org/jdbi/v3/ConstructorMapper.java

```diff
@@ -0,0 +1,174 @@
+
+    @Override
+    public T map(ResultSet rs, StatementContext ctx)
+        throws SQLException
+    {
+        final ResultSetMetaData metadata = rs.getMetaData();
+        final List<String> columnNames = new ArrayList<>(metadata.getColumnCount());
+
+        for (int i = 1; i <= metadata.getColumnCount(); ++i) {
+            columnNames.add(metadata.getColumnLabel(i));
+        }
+
+        ObjectCreator<T> creator = creatorCache.computeIfAbsent(columnNames, c -> createCreator(c, metadata, ctx));
+        return creator.create(rs);
+    }
+
+    private ObjectCreator<T> createCreator(List<String> columnNames, ResultSetMetaData metadata, StatementContext ctx) {
+        final int length;
+        try {
+            length = metadata.getColumnCount();
+        } catch (SQLException e) {
+            throw new ResultSetException("Could not get column count", e, ctx);
+        }
+
+        if (length != constructor.getParameterCount()) {
+            throw new IllegalStateException(length + " columns in result set, but constructor takes " + constructor.getParameterCount());
+        }
+
+        final int[] columnMap = new int[length];
+        final ColumnMapper<?>[] mappers = new ColumnMapper<?>[length];
+
+        for (int i = 0; i < length; i++) {
+            final Type type = constructor.getGenericParameterTypes()[i];
+            final int paramIndex = parameterIndexForColumn(columnNames.get(i));
+
+            if (mappers[paramIndex] != null) {
+                throw new IllegalArgumentException("Column named " + columnNames.get(i) + " maps to multiple parameters");
+            }
+
+            mappers[paramIndex] = ctx.findColumnMapperFor(type).orElseThrow(() ->
+                new IllegalArgumentException("Could not find column mapper for type '" + type + "' of parameter for constructor " + constructor));
+            columnMap[i] = paramIndex;
+        }
+
+        return rs -> {
+            final Object[] params = new Object[length];
+            for (int i = 0; i < length; i++) {
+                params[columnMap[i]] = mappers[i].map(rs, i + 1, ctx);
+            }
+            try {
+                return constructor.newInstance(params);
+            } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+                if (e.getCause() instanceof RuntimeException) {
+                    throw (RuntimeException) e.getCause();
+                }
+                if (e.getCause() instanceof Error) {
+                    throw (Error) e.getCause();
+                }
+                throw new RuntimeException(e);
+            }
+        };
+    }
+
+    private int parameterIndexForColumn(String columnName)
+    {
+        for (int i = 0; i < constructor.getParameterCount(); i++) {
+            for (ColumnNameMappingStrategy strategy : nameMappingStrategies) {
+                if (strategy.nameMatches(paramName(constructor.getParameters()[i]), columnName)) {
+                    return i;
```

Comment:
I think that is dangerous.  Right now if you accidentally do `SELECT b, a` with `MyBean(String a, String b)` it will catch it -- if we just use ordering, we cannot.  Additionally if you do a `SELECT *` I'm not entirely clear that we can or should rely on the column order.